src/wallet/silentpayment.cpp

```diff
@@ -0,0 +1,115 @@
+#include "silentpayment.h"
+
+#include <key_io.h>
+#include <secp256k1.h>
+#include <secp256k1_ecdh.h>
+#include <secp256k1_extrakeys.h>
+
+namespace silentpayment {
+
+secp256k1_pubkey Sender::CreateTweakedPubkey(
+    const secp256k1_context* ctx,
+    const secp256k1_xonly_pubkey recipient_public_key,
+    const unsigned char sender_secret_key[32])
+{
+    unsigned char shared_secret[32];
+
+    secp256k1_pubkey output_pubkey;
+
+    int return_val;
+
+    return_val = secp256k1_ecdh_xonly(ctx, shared_secret, &recipient_public_key, sender_secret_key, NULL, NULL);
+    assert(return_val);
+
+    return_val = secp256k1_ec_seckey_verify(secp256k1_context_no_precomp, shared_secret);
+    assert(return_val);
+
+    return_val = secp256k1_xonly_pubkey_tweak_add(ctx, &output_pubkey, &recipient_public_key, shared_secret);
+    assert(return_val);
+
+    return output_pubkey;
+}
+
+XOnlyPubKey Sender::CreateSilentPaymentAddress(
+    const CKey& sender_secret_key,
+    const XOnlyPubKey& _recipient_public_key)
+{
+    int return_val;
+
+    secp256k1_context* ctx = secp256k1_context_create(SECP256K1_CONTEXT_SIGN | SECP256K1_CONTEXT_VERIFY);
+
+    secp256k1_xonly_pubkey recipient_public_key;
+    return_val = secp256k1_xonly_pubkey_parse(ctx, &recipient_public_key, _recipient_public_key.data());
+    assert(return_val);
+
+    secp256k1_pubkey sender_output_pubkey;
+    sender_output_pubkey = CreateTweakedPubkey(ctx, recipient_public_key, sender_secret_key.data());
+
+    // Test the key
+    size_t len;
+    unsigned char sender_serialized_output_pubkey[33];
+    len = sizeof(sender_serialized_output_pubkey);
+    return_val = secp256k1_ec_pubkey_serialize(ctx, sender_serialized_output_pubkey, &len, &sender_output_pubkey, SECP256K1_EC_COMPRESSED);
+    assert(return_val);
+
+    CPubKey pubKey = CPubKey(sender_serialized_output_pubkey);
+
+    assert(pubKey.IsFullyValid());
+
+    XOnlyPubKey xOnlyPubKey = XOnlyPubKey(pubKey);
+
+    return xOnlyPubKey;
+}
+
+void Recipient::CreateTweakedPubkey(
```

Comment:
I'm just trying to understand the protocol better. This function is called "create tweaked pubkey" but what seems more significant about it to me is that it also provides the recipient with the private key they need to spend from the silent payment output. Is that correct?

Reply:
Yes. I defined this name when using another approach, but that's the idea.

`secp256k1_keypair_xonly_tweak_add` returns both the tweaked private and public keys set in the `recipient_keypair` parameter. So the recipient can spend the output.